src/connections.jl

```diff
@@ -381,12 +398,12 @@ This function calls [`PQfinish`](https://www.postgresql.org/docs/10/libpq-connec
 but only if `jl_conn.closed` is `false`, to avoid a double-free.
 """
 function Base.close(jl_conn::Connection)
-    if !jl_conn.closed
-        libpq_c.PQfinish(jl_conn.conn)
+    if !atomic_cas!(jl_conn.closed, false, true)
+        lock(jl_conn.lock) do
+            libpq_c.PQfinish(jl_conn.conn)
+            jl_conn.conn = C_NULL
+        end
     end
```

Comment:
I think there is a race condition between `close!` and `reset!`
that can leave
`jl_conn.closed[] == false` and `jl_conn.conn == C_NULL`
but I would need to draw it out and it is way to late in the day for me to be checking for race conditions.
So will do it tomorrow unless you can confirm I am right.
(or prove i am wrong)

Reply:
Ok I have looked closely and convinced myself there is no race.
Between `close!` and `reset!`
nor between those functions and themselves.